hystrix-contrib/hystrix-javanica/hystrix-javanica-core/src/main/java/com/netflix/hystrix/contrib/javanica/command/MetaHolder.java

```diff
@@ -15,96 +15,94 @@
-@Immutable
-public final class MetaHolder {
-
-    private final HystrixCollapser hystrixCollapser;
-    private final HystrixCommand hystrixCommand;
-    private final DefaultProperties defaultProperties;
-
-    private final Method method;
-    private final Method cacheKeyMethod;
-    private final Method ajcMethod;
-    private final Method fallbackMethod;
-    private final Object obj;
-    private final Object proxyObj;
-    private final Object[] args;
-    private final Closure closure;
-    private final String defaultGroupKey;
-    private final String defaultCommandKey;
-    private final String defaultCollapserKey;
-    private final String defaultThreadPoolKey;
-    private final ExecutionType executionType;
-    private final boolean extendedFallback;
-    private final ExecutionType collapserExecutionType;
-    private final ExecutionType fallbackExecutionType;
-    private final boolean fallback;
-    private boolean extendedParentFallback;
-    private final boolean defaultFallback;
-    private final JoinPoint joinPoint;
-    private final boolean observable;
-    private final ObservableExecutionMode observableExecutionMode;
-
-    private static final Function identityFun = new Function<Object, Object>() {
-        @Nullable
-        @Override
-        public Object apply(@Nullable Object input) {
-            return input;
-        }
-    };
-
-    private MetaHolder(Builder builder) {
-        this.hystrixCommand = builder.hystrixCommand;
-        this.method = builder.method;
-        this.cacheKeyMethod = builder.cacheKeyMethod;
-        this.fallbackMethod = builder.fallbackMethod;
-        this.ajcMethod = builder.ajcMethod;
-        this.obj = builder.obj;
-        this.proxyObj = builder.proxyObj;
-        this.args = builder.args;
-        this.closure = builder.closure;
-        this.defaultGroupKey = builder.defaultGroupKey;
-        this.defaultCommandKey = builder.defaultCommandKey;
-        this.defaultThreadPoolKey = builder.defaultThreadPoolKey;
-        this.defaultCollapserKey = builder.defaultCollapserKey;
-        this.defaultProperties = builder.defaultProperties;
-        this.hystrixCollapser = builder.hystrixCollapser;
-        this.executionType = builder.executionType;
-        this.collapserExecutionType = builder.collapserExecutionType;
-        this.fallbackExecutionType = builder.fallbackExecutionType;
-        this.joinPoint = builder.joinPoint;
-        this.extendedFallback = builder.extendedFallback;
-        this.defaultFallback = builder.defaultFallback;
-        this.fallback = builder.fallback;
-        this.extendedParentFallback = builder.extendedParentFallback;
-        this.observable = builder.observable;
-        this.observableExecutionMode = builder.observableExecutionMode;
-    }
-
-    public static Builder builder() {
-        return new Builder();
-    }
+public abstract class MetaHolder<M extends MetaHolder<M,B>,B extends MetaHolder.Builder<M, B>> {
```

Comment:
Type Parameters are used at two places in MetaHolder.

1) Constructor
```java
protected  MetaHolder(B builder) {
	this.hystrixCommand = builder.hystrixCommand;
        // other properties here
}
```
2) Copy method

```java
protected abstract B copy();
```
Class level Type parameters can be replaced by method level parameters as shown below.

```java

protected <M extends MetaHolder, B extends MetaHolder.Builder<M, B>> MetaHolder(B builder) {
         this.hystrixCommand = builder.hystrixCommand;
        // other properties here
}
protected abstract <M extends MetaHolder, V extends MetaHolder.Builder<M, V>> V copy();
```
The only down side is that above change will show an `unchecked warning` in classes which implement copy() method. I am not sure which approach(method/class level) is ideal in this case. Let me know your thoughts. `copy()` method is used in `LazyCommandExecutionAction`. Also, do you think there is a better way to handle `copy` operation?

In the builder also, we need to pass `Type` of associated MetaHolder  because following snippet in builder is called from a lot of places and without type info we may not get desired behavior. 

```java
public abstract T build();
```

Replacing above snippet as `public  MetaHolder build() ` will not help as we need to know the exact sub class of MetaHolder that is built at the compile time.